Libraries/LibGfx/SystemTheme.h

```diff
@@ -143,9 +143,9 @@ enum class PathRole {
 };
 
 struct SystemTheme {
-    Color color[(int)ColorRole::__Count];
+    RGBA32 color[(int)ColorRole::__Count];
     int metric[(int)MetricRole::__Count];
-    String path[(int)PathRole::__Count];
+    char path[(int)PathRole::__Count][256]; // TODO: PATH_MAX?
```

Comment:
To be clear, 256 is of course wrong too.

Reply:
That depends on what exactly "trivially copyable" means.

* Is it fine to share a dynamically-sized string buffer in shared memory? Sure.
* If you declare it using a flexible array member, can the compiler trivially memcpy it instead of invoking a copy constructor? No.

But generally, any userspace code that assumes / relies on the fact that paths are shorter than some fixed number of characters (whether 256 or 4096) is wrong / broken, and there's been an effort to clean up our userspace from this assumption :smile: The proper way is to always use dynamically sized buffers, including `AK::String`s where possible.

Reply:
Well, do we have such a thing? We have nothing like `std::allocator` that would allow me to place a `String` into the `SharedBuffer`. Actually, [I do have a branch that adds this concept to AK](https://github.com/tomuta/serenity/commit/3c3bfaf4e2180cbce13c9bbb7034720976432baa), but it's quite stale as I haven't worked on it in a long time.

Again, don't disagree, hence the `TODO`. I would argue this is less of a problem than what we have today: memory corruption.

Reply:
That still wouldn't solve the problems of internal pointers being potentially off, depending on where the buffer ends up being mapped.

Reply:
> That's why a put a TODO there ;-) Right now, the code corrupts memory at best, or crashes. I didn't mean to come up with the perfect solution here. With my other changes in the works, I am unable to boot because of this bug...

I mean, great catch and this surely needs fixing; it's just the `PATH_MAX` that triggers me :slightly_smiling_face: We need to do that bit differently.

Reply:
> Well, do we have such a thing? We have nothing like `std::allocator` that would allow me to place a `String` into the `SharedBuffer`.

Yeah, I'm not suggesting we use a `String` in this case, just a dynamically-sized buffer of some sort.

> I would argue this is less of a problem than what we have today: memory corruption.

Totally, yeah.

Reply:
> Would you volunteer to fix this? ;-)

Sure :slightly_smiling_face: Set me try...

Reply:
@thankyouverycool would it be fine with you if I unimplemented "path roles" and just went with a single `char char title_buttons_path[];`? Because working with a dynamic number of dynamically-sized paths is indeed more complex; we can do that too if needed, but not doing that would be easier.

Reply:
It's actually even simpler than that (and this explains how it has been working up until now): it's only the window server that cares about the title buttons path; that's no need to send the path to the clients at all.

Reply:
> It's actually even simpler than that (and this explains how it has been working up until now): it's only the window server that cares about the title buttons path; that's no need to send the path to the clients at all.

Does this apply to the theme editor, too?

Reply:
The theme editor currently hardcodes title button icon paths 😃
https://github.com/SerenityOS/serenity/blob/00f47bba237bc87c648e0777f0f4ee9eaca9aa4f/Applications/ThemeEditor/PreviewWidget.cpp#L96-L98